### PR TITLE
nixos/systemd: install sysctl snippets

### PR DESCRIPTION
**Pull request by flokli**
systemd provides two sysctl snippets, 50-coredump.conf and
50-default.conf.

Apart from some loose reverse path filtering and source routing
filtering, this enables fq_codel as a packet scheduler to fight
bufferbloat, and configures systemd-coredump properly.

Let's start using these, like other distros already do for quite some
time, and remove those duplicate `boot.kernel.sysctl` options we
previously did set.

In the case of rp_filter (which systemd would set to 2 (loose)), make
our overrides to "1" more explicit.

sysctl.d(5) recommends prefixing all filenames in /etc/sysctl.d with a
two-digit number and a dash, to simplify the ordering of the files.

Some packages provide custom files, often with "50-" prefix.
To ensure user-supplied configuration takes precedence over the one
specified via `boot.kernel.sysctl`, prefix the file generated there with
"60-".

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: nixos/modules/profiles/hardened.nix [hunk 1]
```diff
@@ -92,23 +92,17 @@ with lib;
...
 
-  # Enable reverse path filtering (that is, do not attempt to route packets
-  # that "obviously" do not belong to the iface's network; dropped packets are
-  # logged as martians).
+  # Enable strict reverse path filtering (that is, do not attempt to route
+  # packets that "obviously" do not belong to the iface's network; dropped
+  # packets are logged as martians).
   boot.kernel.sysctl."net.ipv4.conf.all.log_martians" = mkDefault true;
-  boot.kernel.sysctl."net.ipv4.conf.all.rp_filter" = mkDefault true;
+  boot.kernel.sysctl."net.ipv4.conf.all.rp_filter" = mkDefault "1";
```
Comment (andir) [inline nixos/modules/profiles/hardened.nix]: Why do we have to change this? The sysctl module maps `true` to `1` for us (since `toString true == "1"`).

Comment (joachifm) [reply nixos/modules/profiles/hardened.nix]: I wondered the same. I think the answer is

> In the case of rp_filter (which systemd would set to 2 (loose)), make
our overrides to "1" more explicit.

**Comment by andir** [comment]
OK, then it is reasonable IMO.

File: nixos/modules/profiles/hardened.nix (hunk 1 above)
Comment (flokli) [reply nixos/modules/profiles/hardened.nix]: Yes. `systemd` sets these to "2" (loose) in the snippets already, the hardenened rules set it to "1" (strict).

Having a boolean here doesn't really explain what and why we're doing.
I also updated the comment above, to add the "strict" bit.

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 2]
```diff
@@ -403,6 +403,26 @@
...
+      <link linkend="opt-boot.kernel.sysctl">boot.kernel.sysctl</link> was moved to
+      <literal>/etc/sysctl.d/60-nixos.conf</literal>, as <literal>sysctl.d(5)</literal>
+      recommends prefixing all filenames in <literal>/etc/sysctl.d</literal> with a
+      two-digit number and a dash to simplify the ordering of the files.
+    </para>
+   </listitem>
+   <listitem>
+     <para>
+       We now install the sysctl snippets shipped with systemd. Apart from some loose reverse path
+       filtering and source routing filtering, this enables <literal>fq_codel</literal> as a packet
```
Comment (grahamc) [inline nixos/doc/manual/release-notes/rl-1909.xml]: "some loose reverse path filtering and source routing filtering" sounds important. Can you go in to more detail about what changed from (old) to (new), and its impact?

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 3]
```diff
@@ -403,6 +403,26 @@
...
+       scheduler to fight bufferbloat, and configures <literal>systemd-coredump</literal> properly.
```
Comment (grahamc) [inline nixos/doc/manual/release-notes/rl-1909.xml]: I maybe should have left the above comment down a few lines. In general when it comes to release notes and setting changes, avoid [weasel words](https://en.wikipedia.org/wiki/Weasel_word).

Comment (flokli) [reply nixos/doc/manual/release-notes/rl-1909.xml]: About the rp_filter option from above comment: systemd sets `net.ipv4.conf.all.rp_filter` to `2` by default, which does enable some reverse path filtering, while we currently don't do any reverse path filtering at all. On hardened and google-compute-config, we set it to `1` (strict).

Detailled explanation about the options:
 - 0 No source address validation is performed and any packet is forwarded to the destination network
 - 1 Strict Mode as defined in RFC 3074. Each incoming packet to a router is tested against the routing table and if the interface that the packet is received on is not the best return path for the packet then the packet is dropped.
 - 2 Loose mode as defines in RFC 3074 Loose Reverse Path. Each incoming packet is tested against the route table and the packet is dropped if the source address is not routable through any interface. The allows for asymmetric routing where the return path may not be the same as the source path.

As a lot of distros enabled the systemd sysctls, they enabled loose reverse path filtering, so I thought it should be fine to do so too. Same goes for `net.ipv4.conf.all.accept_source_route = 0`.

I didn't want to explicitly list every option currently set in `${systemd}/example/sysctl.d/50-default.conf` - only the fact that we now apply these options, plus some broad summary about what's changed, so people are aware when reading release notes.

**Comment by flokli** [comment]
Regarding systemd-coredump:

We seem to currently have a `systemd.coredump` module, which duplicates the `kernel.core_pattern` from the upstream systemd unit if `systemd.coredump.enable` is set to true, otherwise sets it back to `core`.

I was a bit baffled `coredumpctl` doesn't save coredumps on NixOS by default, and wonder if it's time to merge the things enabled there to the systemd module itself (and remove the module) - people could still get the old behaviour by setting `boot.kernel.sysctl."kernel.core_pattern" = "core"`.

@nh2, @Jookia @auntieNeo, what do you think?

File: nixos/doc/manual/release-notes/rl-1909.xml (hunk 2 above)
Comment (flokli) [reply nixos/doc/manual/release-notes/rl-1909.xml]: see below comment.

**Review by joachifm — APPROVED** [review]

**Comment by nh2** [comment]
As long as we keep the ability to use normal `core` instead of systemd's coredump handling (because it's very slow slow), I'm fine with things being moved around.

**Comment by flokli** [comment]
@nh2 Yes, if you set the sysctl parameter as described in the changelog, systemd-coredump won't be handling coredumps at all, and the kernel will dump the coredump file by itself.

**Comment by flokli** [comment]
I'd be curious to know about when the coredump performance becomes critical though ;-)

**Comment by flokli** [comment]
I removed weaselness in the release notes and explained getting back the old coredump behaviour a bit more clearly in the release notes.

**Comment by nh2** [comment]
> I'd be curious to know about when the coredump performance becomes critical though ;-)

@flokli I have observed many times that when a program with a couple GB resident memory core dumps, systemd's coredump program was spinning on 100% CPU for many minutes, while `core` terminated almost instantly.

Probably it's [this issue](https://lists.freedesktop.org/archives/systemd-devel/2015-April/031357.html).

I didn't investigate it, I just found it very annoying to have to wait so long.

**Comment by flokli** [comment]
This was fixed upstream February 2016 in bdfd7b2c63cac944a3aa1fc0fafd19f41789e208 - anyways, when setting sysctl, it won't be piped through anyways.

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 4]
```diff
@@ -424,6 +424,37 @@
      installer after creating <literal>/var/lib/nextcloud</literal>.
     </para>
    </listitem>
+   <listitem>
+    <para>
+      The <literal>/etc/sysctl.d/nixos.conf</literal> file containing all the options set via
+      <link linkend="opt-boot.kernel.sysctl">boot.kernel.sysctl</link> was moved to
+      <literal>/etc/sysctl.d/60-nixos.conf</literal>, as <literal>sysctl.d(5)</literal>
```
Comment (worldofpeace) [inline nixos/doc/manual/release-notes/rl-1909.xml]:
```suggestion
      <literal>/etc/sysctl.d/60-nixos.conf</literal>, as  <citerefentry><refentrytitle>sysctl.d</refentrytitle<manvolnum>5</manvolnum></citerefentry>
```

File: nixos/doc/manual/release-notes/rl-1909.xml [hunk 5]
```diff
@@ -424,6 +424,37 @@
...
+       scheduler to fight bufferbloat, and configures the kernel to pass coredumps to
+       <literal>systemd-coredump</literal>.
+       These sysctl snippets can be found in <literal>/etc/sysctl.d/50-*.conf</literal>,
+       and overridden via <link linkend="opt-boot.kernel.sysctl">boot.kernel.sysctl</link>
+       (which will place the parameters in <literal>/etc/sysctl.d/60-nixos.conf</literal>).
+     </para>
+   </listitem>
+   <listitem>
+    <para>
+      The <literal>systemd.coredump</literal> module has been merged into the systemd module.
```
Comment (worldofpeace) [inline nixos/doc/manual/release-notes/rl-1909.xml]: I don't think the code changes need to be in the release notes, any information that would be important for end users.

**Comment by worldofpeace** [comment]
I found it impossible to suggest this, but I personally found this part to read pretty like this

```diff
diff --git a/nixos/doc/manual/release-notes/rl-1909.xml b/nixos/doc/manual/release-notes/rl-1909.xml
index f9363561bdc..3ae0492ebfa 100644
--- a/nixos/doc/manual/release-notes/rl-1909.xml
+++ b/nixos/doc/manual/release-notes/rl-1909.xml
@@ -435,13 +435,26 @@
    </listitem>
    <listitem>
      <para>
-       We now install the sysctl snippets shipped with systemd. Apart from enabling loose reverse path
-       filtering and source routing filtering, this enables <literal>fq_codel</literal> as a packet
-       scheduler to fight bufferbloat, and configures the kernel to pass coredumps to
-       <literal>systemd-coredump</literal>.
-       These sysctl snippets can be found in <literal>/etc/sysctl.d/50-*.conf</literal>,
-       and overridden via <link linkend="opt-boot.kernel.sysctl">boot.kernel.sysctl</link>
-       (which will place the parameters in <literal>/etc/sysctl.d/60-nixos.conf</literal>).
+       We now install the sysctl snippets that are shipped with systemd.
+       <itemizedlist>
+        <para>This enables:</para>
+        <listitem>
+          <para>Loose reverse path filtering</para>
+        </listitem>
+        <listitem>
+          <para>Source routing filtering</para>
+        </listitem>
+        <listitem>
+         <para>
+          <literal>fq_codel</literal> as a packet scheduler (this helps to fight bufferbloat)
+         </para>
+        </listitem>
+       </itemizedlist>
+
+       And configures the kernel to pass its coredumps to <literal>systemd-coredump</literal>.
+       The sysctl snippets that are shipped with systemd can be found in <literal>/etc/sysctl.d/50-*conf</literal> and can
+       be overridden with the <link linkend="opt-boot.kernel.sysctl">boot.kernel.sysctl</link> option.
+       Note these overrides still go into <literal>/etc/sysctl.d/60-nixos.conf</literal>.
      </para>
    </listitem>
    <listitem>

```

Screenshot from 2019-08-17 21 40 10

File: nixos/doc/manual/release-notes/rl-1909.xml (hunk 5 above)
Comment (flokli) [reply nixos/doc/manual/release-notes/rl-1909.xml]: Right - I dropped that line.

**Comment by flokli** [comment]
Docs should be good, any objections? ;-)

**Review by worldofpeace — APPROVED** [review]
Docs are good :+1:

**Comment by gebner** [comment]
This change disables all SysRq keys (except sync) by default.  Was this intentional?  In any case, this needs to be documented in the release notes.

**Comment by flokli** [comment]
@gebner I indeed missed that line!

It's part of `${pkgs.systemd}/example/sysctl.d/50-default.conf`:

```
# Use kernel.sysrq = 1 to allow all keys.
# See https://www.kernel.org/doc/html/latest/admin-guide/sysrq.html for a list
# of values and keys.
kernel.sysrq = 16
```

I'll update the release notes accordingly and file a follow-up PR, thanks!

**Pull request merged**